concordium-std/src/impls.rs

```diff
@@ -419,12 +473,145 @@ impl<T: sealed::ContextType> HasCommonData for ExternContext<T> {
+                                current_position: 0,
+                            },
+                        })
+                    } else {
+                        crate::trap() // host precondition violation.
+                    }
+                }
+            }
+            0x01 => Err(InvokeError::AmountTooLarge),
+            0x02 => Err(InvokeError::MissingAccount),
+            0x03 => Err(InvokeError::MissingContract),
+            0x04 => Err(InvokeError::MissingEntrypoint),
+            0x05 => Err(InvokeError::MessageFailed),
+            0x06 => Err(InvokeError::Trap),
+            _ => Err(InvokeError::Unknown), // FIXME: This should be trap as well.
+        }
+    }
+}
+
+impl<State: Deserial + Serial> HasHost<State> for Host<State> {
+    type CallResponseType = CallResponse;
+
+    fn invoke_transfer(&mut self, receiver: &AccountAddress, amount: Amount) -> InvokeResult<()> {
+        let mut bytes: MaybeUninit<[u8; ACCOUNT_ADDRESS_SIZE + 8]> = MaybeUninit::uninit();
+        let data = unsafe {
+            (bytes.as_mut_ptr() as *mut u8).copy_from_nonoverlapping(
+                receiver.as_ref() as *const [u8; ACCOUNT_ADDRESS_SIZE] as *const u8,
+                ACCOUNT_ADDRESS_SIZE,
+            );
+            (bytes.as_mut_ptr() as *mut u8).add(ACCOUNT_ADDRESS_SIZE).copy_from_nonoverlapping(
+                &amount.micro_ccd.to_le_bytes() as *const [u8; 8] as *const u8,
+                8,
+            );
+            bytes.assume_init()
+        };
+        let response = unsafe {
+            invoke(INVOKE_TRANSFER_TAG, data.as_ptr(), (ACCOUNT_ADDRESS_SIZE + 8) as u32)
+        };
+        parse_response_code(response).map(|_| ())
+    }
+
+    fn invoke_contract(
+        &mut self,
+        to: &ContractAddress,
+        parameter: Parameter,
+        method: EntrypointName,
+        amount: Amount,
+    ) -> InvokeResult<(bool, Option<Self::CallResponseType>)> {
+        // calling V0 contracts returns None
+        let mut data =
+            Vec::with_capacity(16 + parameter.0.len() + 2 + method.size() as usize + 2 + 8);
+        let mut cursor = Cursor::new(&mut data);
+        to.serial(&mut cursor).unwrap_abort();
+        parameter.serial(&mut cursor).unwrap_abort();
+        method.serial(&mut cursor).unwrap_abort();
+        amount.serial(&mut cursor).unwrap_abort();
+        // serialize state since it might have been modified
+        // FIXME: Only do this if needed.
+        self.state.serial(&mut ContractState::open(())).unwrap_abort();
+        let len = data.len();
+        let response = unsafe { invoke(INVOKE_CALL_TAG, data.as_ptr(), len as u32) };
+        let (state_modified, res) = parse_response_code(response)?;
+        if state_modified {
+            // The state of the contract changed as a result of the call.
+            // So we refresh it.
+            if let Ok(new_state) = (&mut ContractState::open(())).get() {
```

Comment:
This does not work for `HasHost<ContractState>` which is what `low_level` methods will use.
Do we still intend to support the low_level interface in v1?

Reply:
One, less than ideal, option is: 
- Implement `Serialize` for `ContractState`, where `serial` writes nothing and `deserial` always resets the cursor to `0`.